[Wu Zihao] iP

zihaowrez opened the pull request:
# Duke v0.1
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Duke frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ *SUPER* FAST to use

All you need to do is,

1. download it from [here](https://github.com/zihaowrez/CS2103_iP/releases/tag/A-Jar).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you :wink:

And it is __FREE__!

Features:

- [x] Managing tasks with/without time limits
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Commit:
Level-1

Commit:
Level-2

Commit:
Level-3

Commit:
Level-4

Commit:
A-TextUiTesting

Commit:
Level-5

Commit:
Level-6

Commit:
Add newline to end of file

Commit:
Level-7

Commit:
Level-8

Commit:
Add a commit to the master branch in order to merge branches Level-7 and Level-8

Commit:
Merge branch 'branch-Level-7'

# Conflicts:
#	src/main/java/Duke.java

Commit:
Merge branch 'branch-Level-8'

Commit:
Delete extra lines due to merging

Commit:
A-MoreOOP

Commit:
A-Packages

Commit:
Add JUnit test for class DukeDateTime

Commit:
Add JavaDoc to every method. Use enums to handle icons.

Commit:
A-CodingStandard

Commit:
Level-9

Commit:
Move class TaskList into package duke.task

Commit:
Merge branch 'add-gradle-support' of https://github.com/zihaowrez/cs2103_ip into branch-A-Gradle

Commit:
A-Gradle

Commit:
Problematic implementation of GUI with controller

Commit:
GUI v0.1

tanweien commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,84 @@
+        if (!records.createNewFile()) {
+            Scanner sc_file = new Scanner(records);
+            while (sc_file.hasNext()) {
+                String[] record = sc_file.nextLine().split(" ", 3);
+                switch (record[0]) {
+                case "T":
+                    tasks.add(new Todo(record[2]));
+                    break;
+                case "D":
+                    String[] desc_by = record[2].split(" /by ", 2);
+                    DukeDateTime by = DukeDateTime.parse(desc_by[1]);
+                    tasks.add(new Deadline(desc_by[0], by));
+                    break;
+                case "E":
+                    String[] desc_at = record[2].split(" /at ", 2);
+                    DukeDateTime at = DukeDateTime.parse(desc_at[1]);
+                    tasks.add(new Event(desc_at[0], at));
+                    break;
+                }
+                if (record[1].equals("1")) {
+                    tasks.set(tasks.size() - 1, tasks.get(tasks.size() - 1).mark());
```
Overall very neat code, however you tend to use magic numbers so maybe consider removing them and following checkstyle.

tanweien commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,75 @@
+
+public class TaskList {
+
+    private final ArrayList<Task> list;
```
Nice implementation of TaskList that follows immutability of OOP convention

tanweien commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,100 @@
+    }
+
+    protected final String description;
+    protected final boolean isDone;
```
Appropriate boolean naming. Good job!

tanweien commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,60 @@
+            }
+            return new TodoCommand(remaining);
+        } else if (firstWord.equals("deadline")) {
+            String[] desc_by = remaining.split(" /by ", 2);
+            if (desc_by.length < 2) {
+                throw new DukeException("Must specify a deadline");
+            } else if (desc_by[0].length() == 0) {
+                throw new DukeException("The description of a deadline cannot be empty");
+            }
+            DukeDateTime datetime = DukeDateTime.parse(desc_by[1]);
+            return new DeadlineCommand(desc_by[0], datetime);
+        } else if (firstWord.equals("event")) {
+            String[] desc_at = remaining.split(" /at ", 2);
+            if (desc_at.length < 2) {
+                throw new DukeException("Must specify the time of the event");
+            } else if (desc_at[0].length() == 0) {
+                throw new DukeException("The description of an event cannot be empty");
+            }
+            DukeDateTime datetime = DukeDateTime.parse(desc_at[1]);
+            return new EventCommand(desc_at[0], datetime);
```
If possible try to avoid using magic numbers

teominghuang commented on src/main/java/duke/Storage.java:
```diff
@@ -50,6 +61,11 @@ public TaskList load() throws IOException, DukeException {
+     * @throws IOException if an I/O error occurred
+     */
     public void update(TaskList tasks) throws IOException {
         File records = new File(filePath + "/data/records.txt");
```
A slight tendency to misname variables and use plural for singular objects, which can be misleading and is discouraged in the textbook. However, still very neat thus far :).

teominghuang commented on src/main/java/duke/command/EventCommand.java:
```diff
@@ -10,17 +10,28 @@
 
     private final String description;
-    private final DukeDateTime by;
+    private final DukeDateTime at;
```
Very good choice abstracting the concept of DateTime and creating an association class for that!

teominghuang commented on src/main/java/duke/TaskList.java:
```diff
@@ -8,36 +8,68 @@
+     * @param t the new task to add
+     * @return a new list of tasks with the new task added
+     */
     public TaskList add(Task t) {
```
There's a slight tendency in picking unintuitive names, which is fine in small doses, but can be tiring when its code base with thousands of lines.

teominghuang commented on src/main/java/duke/Parser.java:
```diff
@@ -36,6 +27,8 @@ public static Command parse(String input) throws DukeException {
             return new MarkCommand(Integer.parseInt(remaining));
         } else if (input.matches("unmark [1-9]+\\d*")) {
             return new UnmarkCommand(Integer.parseInt(remaining));
+        } else if (firstWord.equals("find")) {
```
Might want to consider using enum to replace the String values to make it neater and easier to manage!

teominghuang commented on src/main/java/duke/Parser.java:
```diff
@@ -36,6 +27,8 @@ public static Command parse(String input) throws DukeException {
             return new UnmarkCommand(Integer.parseInt(remaining));
+        } else if (firstWord.equals("find")) {
+            return new FindCommand(remaining);
         } else if (firstWord.equals("todo")) {
```
Might want to consider using a switch statement instead of an if-else statement. The organised nature of switch-statements make collaboration between teammates easier. It looks neater as well :).

teominghuang reviewed (changes requested):
A lot of thought that were put into the code-base. The hours spent on this is very apparent. The structure was very well-done and it look little time for me to grasp the code base. There are several deep nestings that can be replaced with switch-statements to look neater. Enums can also be considered. Overall, very well done! Just some tiny tweaks here and there to amend.

Commit:
Improve GUI

1. Add paddings around dialog boxes
2. Change font and background color
3. Anchor children of ScrollPane so that the layout changes when the
application window is resized
4. Add prompt text in the input field

Commit:
Delete unused imports in Duke; Add numbers when listing tasks

Commit:
Add assertion for storage path

Commit:
Improve code quality; Delete unused imports and variables

Commit:
Add assertion message

Commit:
Add a .yml file into [root]./workflow

Commit:
Rename folder "workflow" to "workflows"

zihaowrez committed:
Create gradle.yml

zihaowrez committed:
Merge pull request #1 from zihaowrez/branch-A-Assertions

Assert Existence of Storage Folder

Commit:
Merge branch 'master' into branch-A-CodeQuality

Commit:
Merge branch 'branch-A-CI' into branch-A-CodeQuality

zihaowrez committed:
Delete gradle.yml

Commit:
GUI v0.2

Commit:
Add Edit function

Commit:
Move GUI related classes into duke.gui package; minor changes

Commit:
Add HelpCommand; Minor changes; Add JavaDoc for non-inheritance methods

zihaowrez committed:
Update README.md

zihaowrez committed:
Update README.md

zihaowrez committed:
Update README.md

zihaowrez committed:
Update README.md

zihaowrez committed:
Update README.md

zihaowrez committed:
Update README.md

zihaowrez committed:
Update README.md

Commit:
Add a String parameter to Launcher.main() method so that the jar file
can run

Commit:
Fix a few bugs

zihaowrez committed:
Update README.md

zihaowrez committed:
Update README.md

zihaowrez committed:
Update README.md

Commit:
Improve Find function and some error messages

zihaowrez committed:
Set theme jekyll-theme-architect

zihaowrez committed:
Update README.md

zihaowrez committed:
Update README.md

zihaowrez committed:
Update README.md

Commit:
Update UI screenshot

Commit:
Add checkstyle

Commit:
Update Ui.png

zihaowrez committed:
Update README.md

zihaowrez committed:
Update README.md

zihaowrez committed:
Update README.md

zihaowrez committed:
Update README.md

zihaowrez committed:
Update README.md

zihaowrez committed:
Update README.md
